### **Prediction** — a block-group crime model that beats the incumbent *inside* cities

The deployed **agency-scale** model (`04_bg_comparison`) levels whole cities well but fits weakly
*within* a city — it can't tell a city's dangerous blocks from its safe ones. This notebook builds a
block-group model on **`log(weighted rate)`** (absolute, no demeaning) and shows it closes that gap.

**Order of the story:**
1. **The incumbent's block-group problem** — weak block-group fit, for *both* property and total crime
   (Detroit is the sharpest case), taken as the reference row from `04`.
2. **The refresh on PROPERTY crime (9 cities)** first — the wider sample makes every number less
   brittle and shows *sample size is the main lever*.
3. **The same refresh on TOTAL crime (5 cities).**
4. **Per-city breakdown, hyperparameter tuning, and next steps.**

Every configuration is scored on two axes, the same ones `04` used to indict the incumbent:

| axis | question | metrics |
| --- | --- | --- |
| **CALIBRATION** | is each block group placed at the right crime rate? | `r2_oos`, `mae` |
| **RANK** (within-city) | inside a city, are the dangerous blocks flagged? | `recall@10` (headline), `recall@25`, `called_safe` |

- **Splits:** stratified **80/20** (interpolation — a city we've already labeled) and **LOCO**
  (leave-one-city-out — a brand-new city, the honest national-rollout estimate). The gap between them
  is the value of having seen a city before.
- **Predictor ladder:** `ours` → `ours+approved` (de-duplicated incumbent features) →
  `ours+approved+agency` (adds last year's agency-wide crime total). No race / protected predictors.
- **Models:** `RidgeCV` and `LightGBM`, same ladder.


In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

import contextlib
import io
import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import optuna

_here = Path.cwd().resolve()
_root = next(p for p in [_here, *_here.parents] if (p / "src").is_dir())
sys.path.insert(0, str(_root / "src"))

from crime_blockgroup_mapping.constants import CITIES
from regression_modelling.constants import AGENCY_ANCHOR_COL
from regression_modelling.models import diagnostics, incumbent, training
from regression_modelling.models.dataset import (
    load_pooled_table, predictor_set, target_pool,
)
from regression_modelling.models.metrics import fold_metrics, within_city_recall

pd.set_option("display.width", 240)
pd.set_option("display.max_columns", 40)
warnings.filterwarnings("ignore", category=RuntimeWarning)
warnings.filterwarnings("ignore", message="An input array is constant")
optuna.logging.set_verbosity(optuna.logging.WARNING)

In [2]:
# Target pools: the cities each target is actually modelled on (ADR 0008).
# target_pool() applies BOTH tests -- source capability (can this feed support the target?)
# and observed usability (did the extract actually produce crime?) -- and logs every
# exclusion. It is the single owner of city eligibility; do not hardcode pools here.
full_cities = target_pool("wtotal")
all_cities = target_pool("wprop")

# Pooled block-group tables: one row per BG with predictors, observed rates, and the agency total.
pool5 = load_pooled_table(full_cities)
pool10 = load_pooled_table(all_cities)

print("\npool5:", pool5.shape, "| pool10:", pool10.shape, "| wprop cities:", len(all_cities))


target_pool('wtotal'): 5 cities
  excluded 5 (source lacks violent-crime coords): sacramento, san_francisco, pittsburgh, columbus, jacksonville
target_pool('wprop'): 9 cities
  excluded 1 (all-zero crime extract, upstream gap): columbus
  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  kansas_city         473 BGs
  detroit             622 BGs


  pooled             5314 BGs across 5 cities
  dropped 22 zero/NaN-pop BGs -> 5292 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 8 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 5284 remain
  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  sacramento          366 BGs
  san_francisco       679 BGs
  pittsburgh          317 BGs
  kansas_city         473 BGs
  jacksonville        550 BGs
  detroit             622 BGs
  pooled             7226 BGs across 9 cities
  dropped 32 zero/NaN-pop BGs -> 7194 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 12 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 7182 remain

pool5: (5284, 113) | pool10: (7182, 113) | wprop cities: 9


### Benchmark & metric glossary

The `existing (04)` row is the deployed **agency-scale** model, scored with the **same helpers** as
the refresh (via `baseline.existing_model_run`) so the comparison is apples-to-apples. Its numbers are
its own block-group fit from `04_bg_comparison`, computed live for **both** targets — wtotal / 5 cities
and wprop / 9 cities.

All metrics are on the absolute `log1p(weighted rate)` scale, pooled across held-out block groups:

- **`r2_oos`** — pooled out-of-sample R² = 1 − SSE/SST. The headline calibration number. The incumbent
  scores **0.16** (total) / **0.06** (property) here — it predicts a rate but a miscalibrated one; the
  refresh predicts the rate directly and should beat it.
- **`mae`** — mean absolute error on the same log-rate scale.
- **`recall@10`** — the same symmetric overlap on the **worst 10%**: of a city's most-dangerous decile
  by observed rate, the share the model also ranks in its own within-city worst 10%. The **headline rank
  metric** (see the note below).
- **`recall@25`** — a symmetric within-city quartile overlap: of a city's truly-dangerous blocks (its
  **worst 25%** by observed rate), the share the model also ranks in its **own** within-city worst 25%
  by predicted rate (higher = better). Observed top-25 in, predicted top-25 out.
- **`called_safe`** — share of dangerous blocks buried below the model's within-city median (the costly
  miss; lower = better).

**Which recall to report — prefer `recall@10`.** Each `recall@N` has a *chance baseline of N* (a random
model lands an N share of its predictions in its own top-N), so the honest score is the lift over that
floor. `recall@10`'s floor is **0.10** vs `recall@25`'s **0.25**: the refresh's ~0.39 (total) / ~0.36
(property) sits at **~3.9× / ~3.5× chance**, while the incumbent's 0.285 / 0.251 is ~2.9× / ~2.5× — a
clear ~+35% relative gap on the product-relevant extreme-danger tail. On `recall@25` the wide net
flatters everyone (refresh ~0.56 is only ~2.2× chance, incumbent ~0.51 ~2.0×), compressing the very gap
we care about. We therefore **headline `recall@10`** and keep `recall@25` as the wider-net cross-check.

The incumbent row is a single **deployed** number (it has seen every city); every refresh row is honest
**LOCO** (city fully held out) or optimistic **80/20** — so *beating the incumbent under LOCO is the
strong result*.


In [3]:
_POOL = {"wtotal": pool5, "wprop": pool10}
_RUNS = {}


def fit_run(category, pset, split, model="ridge", gbm_params=None):
    """Fit one absolute log-rate run: predictor set x fold protocol x estimator.

    A thin adapter over the single training orchestrator -- the notebook's axes map
    straight onto its arguments, so no spec object is needed. Fold models are retained
    so the diagnostics cells below read the SAME runs rather than refitting them.
    """
    return training.run_cv(
        _POOL[category], category, mode="lograte",
        predictors=predictor_set(pset, category),
        estimator=model, split="loco" if split == "loco" else "holdout",
        gbm_params=gbm_params, keep_fits=True, verbose=False,
    )


def get_run(category, pset, split, model="ridge"):
    """Return a cached absolute log-rate run for one (predictor set, split, model).

    Runs are memoised in ``_RUNS`` so each configuration is fit once, and fitter stdout is
    suppressed to keep tables clean. ``pset`` is 'ours' | 'ours+approved' |
    'ours+approved+agency'; ``split`` is 'loco' | '80_20'; ``model`` is 'ridge' | 'gbm'.
    """
    key = (category, pset, split, model)
    if key not in _RUNS:
        _RUNS[key] = fit_run(category, pset, split, model)
    return _RUNS[key]


def pooled_r2_mae(run, category):
    """Pooled out-of-sample R2 and MAE on the log1p(rate) scale (same recipe as 04_bg_comparison):
    y = log1p(observed rate), yhat = the run's y_pred (already log-rate), over every held-out BG."""
    g = run.scored
    y = np.log1p(g[f"{category}_rate"].astype(float))
    yhat = g["y_pred"].astype(float)
    resid = y - yhat
    r2 = 1 - (resid ** 2).sum() / ((y - y.mean()) ** 2).sum()
    return round(float(r2), 3), round(float(resid.abs().mean()), 3)


def rank_metrics(run, category):
    """Symmetric within-city recall at two nets + called_safe (the costly miss).

    recall@N = of a city's observed worst-N share of BGs, the share the model also ranks in its OWN
    predicted worst-N (chance baseline = N, so read the lift OVER N). recall@10 isolates the
    extreme-danger tail -- the blocks a crime product actually acts on, and where the refresh's lift
    over the incumbent is largest (~4x vs ~2.9x chance); recall@25 is a wider, easier net. called_safe
    = share of the observed worst-25% the model buries below its within-city median (lower = better).
    """
    r10 = within_city_recall(run, category, danger_top=0.10, nets=(0.10,)).loc["POOLED"]
    r25 = within_city_recall(run, category, danger_top=0.25, nets=(0.25,)).loc["POOLED"]
    return {"recall@10": r10["recall@top10"], "recall@25": r25["recall@top25"],
            "called_safe": r25["called_safe"]}


def incumbent_run(category, cities):
    """Deployed incumbent (agency-scale, 04_bg_comparison) as a FoldRun."""
    with contextlib.redirect_stdout(io.StringIO()):
        return incumbent.existing_model_run(category, cities=cities)


def incumbent_row(category, cities):
    """Score the incumbent with the SAME helpers as the refresh -> its reference row.

    ``total_pt_ct`` / ``property_pt_ct`` IS a predicted weighted crime rate, so it carries a real
    (if weak) ``r2_oos`` / ``mae`` -- the as-served block-group fit reported in 04 -- alongside the
    within-city ``recall@10`` / ``recall@25`` / ``called_safe``.
    """
    run = incumbent_run(category, cities)
    r2, mae = pooled_r2_mae(run, category)
    return {"r2_oos": r2, "mae": mae, **rank_metrics(run, category)}


# Incumbent reference rows, computed live for both targets (wtotal / 5, wprop / 9 cities).
EXISTING_WTOTAL = incumbent_row("wtotal", full_cities)
EXISTING_WPROP = incumbent_row("wprop", all_cities)


def combined_table(category, sets, model="ridge", existing=None):
    """Calibration + rank scorecard for one model across predictor sets, LOCO and 80/20.

    CALIBRATION columns (block placed at the right rate): r2_oos, mae.
    RANK columns (dangerous-block flagging within a city): recall@10, recall@25, called_safe.
    An optional ``existing`` dict prepends the deployed incumbent reference row (its own 04 fit).
    """
    rows = []
    if existing is not None:
        rows.append({"predictors": "existing (04)", "split": "deployed", **existing})
    # One block per predictor set; within each, the honest LOCO row then the optimistic 80/20 row.
    for name in sets:
        for split in ["loco", "80_20"]:
            run = get_run(category, name, split, model)
            r2, mae = pooled_r2_mae(run, category)
            rows.append({"predictors": name, "split": split.replace("_", "/"),
                         "r2_oos": r2, "mae": mae, **rank_metrics(run, category)})
    return pd.DataFrame(rows).set_index(["predictors", "split"])


def level_by_city(category, pset, model="ridge", split="loco"):
    """Per-held-out-city decomposition of the pooled r2_oos -- is it uniform or sunk by one city?

    For each held-out city: its observed vs predicted city-MEAN log-rate (`level_err`, the
    calibration miss the agency total is meant to fix) and its own out-of-sample `r2_oos`. A single
    strongly-negative city is what drags the pooled number down (the small-sample risk on wtotal/5).
    """
    run = get_run(category, pset, split, model)
    g = run.scored.copy()
    g["_obs"] = np.log1p(g[f"{category}_rate"].astype(float))
    rows = []
    # One row per held-out city: its level miss and its own r2_oos.
    for city, d in g.groupby("holdout_city"):
        y, p = d["_obs"], d["y_pred"]
        r2 = 1 - ((y - p) ** 2).sum() / ((y - y.mean()) ** 2).sum()
        rows.append({"city": city, "n": len(d),
                     "obs_mean": round(float(y.mean()), 2), "pred_mean": round(float(p.mean()), 2),
                     "level_err": round(float(p.mean() - y.mean()), 2), "r2_oos": round(float(r2), 3)})
    return pd.DataFrame(rows).set_index("city").sort_values("r2_oos")


### Predictor sets

Exactly what each rung feeds the model. The ladder is additive: `ours+approved` adds the approved,
de-duplicated incumbent features to `ours`; `ours+approved+agency` adds **one** column,
`agency_lag_<target>_log` — last year's agency-wide crime total, identical across every BG in a city
(so it carries the city-level variance; the BG predictors carry the rest).

In [4]:
# Print the resolved predictor columns for each rung (wtotal pool), showing what each rung adds.
_seen = []
for pset in ["ours", "ours+approved", "ours+approved+agency"]:
    cols = [c for c in predictor_set(pset, "wtotal") if c in pool5.columns]
    added = [c for c in cols if c not in _seen]      # only the columns this rung introduces
    print(f"{pset}  (total {len(cols)}, +{len(added)} new):")
    for c in added:
        print("   +", c)
    _seen = cols
    print()

ours  (total 23, +23 new):
   + moved1yr_pct
   + own_pct
   + lap_pct
   + city_centers_dist
   + pop_est_5mile_log
   + pop_ch_1mile
   + vacant_pct_log
   + clip_liens_pct_log
   + clip_foreclosure_pct_log
   + unq_convenience_stores_clips
   + unq_gas_stations_clips
   + unq_liquor_stores_clips
   + vacant_pct_lag6_log
   + clip_liens_pct_lag6_log
   + clip_foreclosure_pct_lag6_log
   + transit_has_transit
   + transit_service_intensity_logc
   + transit_nearest_stop_m_log
   + transit_overnight_stop_share
   + roof_condition_avg
   + roof_debris_pct_avg
   + hardscapes_pct_avg
   + roof_missing_material_pct

ours+approved  (total 28, +5 new):
   + in_household_pct
   + det_pct
   + div_encentral
   + div_midatlantic
   + div_southatlantic

ours+approved+agency  (total 29, +1 new):
   + agency_lag_wtotal_log



### 1 · The incumbent's block-group problem — property *and* total

The deployed agency-scale model fits block-group crime **weakly**, on both crime families (the full
per-city breakdown lives in `04_bg_comparison`; here it is the reference row the refresh must beat):

- **Calibration is poor.** Pooled out-of-sample `r2_oos` is **0.16** (total) / **0.06** (property) on
  the log-rate scale — and several cities go **negative** (the served prediction underperforms a
  per-city constant). **Detroit** is the sharpest case.
- **Within-city ranking is only fair.** `recall@10` ≈ **0.29** (total) / **0.25** (property) — barely
  above the 0.10 chance floor on the extreme-danger tail — with `recall@25` ≈ 0.51 / 0.48 and
  `called_safe` ≈ 0.21 / 0.24 (dangerous blocks it buries below the city median).

That block-group gap is the whole target of the refresh.


In [5]:
# The incumbent's own block-group fit -- the reference row for every table below (both targets).
print("INCUMBENT (deployed agency-scale) reference -- its own block-group fit, from 04_bg_comparison\n")
print("  wtotal / 5 cities:", EXISTING_WTOTAL)
print("  wprop  / 9 cities:", EXISTING_WPROP)
print("\nWeak calibration (r2_oos 0.16 / 0.06; several cities negative, Detroit worst) and only fair")
print("within-city ranking. See 04_bg_comparison for the per-city detail; the refresh below closes it.")

INCUMBENT (deployed agency-scale) reference -- its own block-group fit, from 04_bg_comparison

  wtotal / 5 cities: {'r2_oos': 0.164, 'mae': 0.787, 'recall@10': np.float64(0.285), 'recall@25': np.float64(0.513), 'called_safe': np.float64(0.208)}
  wprop  / 9 cities: {'r2_oos': 0.058, 'mae': 0.712, 'recall@10': np.float64(0.251), 'recall@25': np.float64(0.475), 'called_safe': np.float64(0.239)}

Weak calibration (r2_oos 0.16 / 0.06; several cities negative, Detroit worst) and only fair
within-city ranking. See 04_bg_comparison for the per-city detail; the refresh below closes it.


### 2 · The refresh on PROPERTY crime (9 cities)

We lead with **property** because it has the **wider sample** (9 cities vs 5), which makes every
cross-city number less brittle — the clearest way to see that **sample size, not model cleverness, is
the main lever.** Same predictor ladder (`ours → +approved → +agency`), both models, scored on LEVEL
and RANK against the incumbent benchmark from section 1.

Read the two splits as a pair: **80/20** (a city we've already labeled — interpolation) vs **LOCO** (a
brand-new city — extrapolation). Watch the agency lagged total lift cross-city LEVEL, and watch the
80/20→LOCO gap as the price of never having seen a city.


In [6]:
SETS = ["ours", "ours+approved", "ours+approved+agency"]
print(f"BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · RidgeCV")
print(combined_table("wprop", SETS, model="ridge", existing=EXISTING_WPROP).to_string())
print("\n* existing (04) is the deployed model's own block-group fit (its reference row).")
print("  ours / +approved / +agency rows are honest LOCO and optimistic 80/20.")

BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / 9 cities · RidgeCV


                               r2_oos    mae  recall@10  recall@25  called_safe
predictors           split                                                     
existing (04)        deployed   0.058  0.712      0.251      0.475        0.239
ours                 loco       0.327  0.595      0.339      0.543        0.188
                     80/20      0.385  0.571      0.369      0.542        0.151
ours+approved        loco       0.358  0.579      0.352      0.538        0.186
                     80/20      0.417  0.557      0.411      0.542        0.159
ours+approved+agency loco       0.380  0.566      0.357      0.543        0.190
                     80/20      0.422  0.554      0.390      0.536        0.159

* existing (04) is the deployed model's own block-group fit (its reference row).
  ours / +approved / +agency rows are honest LOCO and optimistic 80/20.


In [7]:
print(f"BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · LightGBM")
print(combined_table("wprop", SETS, model="gbm", existing=EXISTING_WPROP).to_string())
print("\n* same predictor ladder; LightGBM replaces RidgeCV. Compare LOCO rows head-to-head.")

BLOCK-GROUP FIT · absolute log(weighted PROPERTY rate) · wprop / 9 cities · LightGBM


                               r2_oos    mae  recall@10  recall@25  called_safe
predictors           split                                                     
existing (04)        deployed   0.058  0.712      0.251      0.475        0.239
ours                 loco       0.270  0.633      0.342      0.513        0.200
                     80/20      0.460  0.537      0.348      0.554        0.165
ours+approved        loco       0.360  0.592      0.349      0.511        0.195
                     80/20      0.483  0.527      0.397      0.554        0.154
ours+approved+agency loco       0.359  0.594      0.367      0.513        0.199
                     80/20      0.490  0.521      0.390      0.557        0.157

* same predictor ladder; LightGBM replaces RidgeCV. Compare LOCO rows head-to-head.


**Read — property refresh (9 cities).**

- **The refresh calibrates block-group crime the incumbent can't.** RidgeCV LOCO `r2_oos` climbs
  **0.33 → 0.36 → 0.38** across the ladder — against the incumbent's **0.06**. We predict a genuine
  per-BG rate; the incumbent's served rate barely beats a constant.
- **We beat it on rank too, under honest LOCO.** `recall@10` ≈ **0.35 vs 0.25** (the headline tail
  metric — ~3.5× vs ~2.5× the 0.10 chance floor), `recall@25` ≈ 0.54 vs **0.48**, `called_safe`
  ≈ 0.19 vs **0.24** (fewer of the worst blocks buried below the city median).
- **The agency lagged total is a decisive calibration predictor.** It lifts Ridge LOCO `r2_oos`
  0.33 → 0.38 (80/20 0.39 → 0.42) — it carries the city-level rate every BG in a city shares, so the
  BG predictors are freed to carry the rest.
- **Sample size is the main lever, not model cleverness.** Every 80/20 `r2_oos` beats its LOCO twin
  (e.g. `ours` Ridge 0.39 vs 0.33) — the gap is purely "have we labeled this city," and it is the
  biggest swing on the page.
- **Ridge vs GBM: Ridge extrapolates more stably.** At the top rung Ridge LOCO `r2_oos` **0.380** > GBM
  **0.359**, and GBM's 80/20→LOCO drop is far larger (0.490 → 0.359 = 0.13) than Ridge's
  (0.422 → 0.380 = 0.04) — GBM overfits the cities it has seen. On rank the two are a near-tie.


### 3 · The same refresh on TOTAL crime (5 cities)

Weighted-**total** covers only the 5 full-coverage cities, so its cross-city calibration rests on 5
city levels — powerful but fragile (the per-city breakdown below opens it up). Same ladder, both
models, against the wtotal incumbent reference row.


In [8]:
print("BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV")
print(combined_table("wtotal", SETS, model="ridge", existing=EXISTING_WTOTAL).to_string())
print("\n* existing (04) is the deployed model's own block-group fit. Refresh rows are honest LOCO and 80/20.")

BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV


                               r2_oos    mae  recall@10  recall@25  called_safe
predictors           split                                                     
existing (04)        deployed   0.164  0.787      0.285      0.513        0.208
ours                 loco       0.245  0.722      0.421      0.592        0.147
                     80/20      0.502  0.599      0.304      0.587        0.131
ours+approved        loco       0.232  0.718      0.409      0.589        0.148
                     80/20      0.514  0.594      0.314      0.587        0.143
ours+approved+agency loco       0.194  0.751      0.411      0.587        0.153
                     80/20      0.515  0.593      0.324      0.591        0.139

* existing (04) is the deployed model's own block-group fit. Refresh rows are honest LOCO and 80/20.


In [9]:
print("BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 cities · LightGBM")
print(combined_table("wtotal", SETS, model="gbm", existing=EXISTING_WTOTAL).to_string())
print("\n* same predictor ladder as the RidgeCV table; compare LOCO rows head-to-head.")

BLOCK-GROUP FIT · absolute log(weighted rate) · wtotal / 5 cities · LightGBM


                               r2_oos    mae  recall@10  recall@25  called_safe
predictors           split                                                     
existing (04)        deployed   0.164  0.787      0.285      0.513        0.208
ours                 loco       0.353  0.696      0.363      0.556        0.176
                     80/20      0.516  0.604      0.333      0.556        0.139
ours+approved        loco       0.379  0.679      0.393      0.564        0.172
                     80/20      0.521  0.601      0.392      0.571        0.143
ours+approved+agency loco       0.375  0.686      0.397      0.558        0.170
                     80/20      0.529  0.596      0.363      0.583        0.151

* same predictor ladder as the RidgeCV table; compare LOCO rows head-to-head.


**Read — total refresh (5 cities).**

- **Same rank win over the incumbent.** LOCO `recall@10` ≈ **0.39 vs 0.29** (the headline tail metric),
  `recall@25` ≈ 0.56 vs **0.51**, `called_safe` ≈ 0.17 vs **0.21** (fewer costly misses). This is the
  Detroit-style within-city gap being closed.
- **Calibration beats the incumbent but is fragile on 5 cities.** GBM LOCO `r2_oos` ≈ 0.35–0.38 (vs
  incumbent **0.16**) and holds across the ladder. RidgeCV LOCO `r2_oos` instead *falls*
  0.245 → 0.232 → **0.194** as the agency total is added — the opposite of property — because on only 5
  cities that cross-city signal is brittle (the per-city breakdown shows one city, Atlanta, causes it).
  Both 80/20 rows sit at ~0.50–0.53.
- **The Ridge-vs-GBM verdict flips vs property** (GBM wins calibration on wtotal/5, Ridge on wprop/9) —
  a small-sample artifact, not a real skill gap. Judge extrapolation on the 9-city property evidence.
- **Both models beat the incumbent on rank under LOCO on both targets** — the robust, repeatable result.


### 4 · Per-city breakdown — is the pooled `r2_oos` uniform or carried by 1–2 cities?

Pooled `r2_oos` is a single fit over every held-out block group, so one badly-calibrated city can
dominate it — especially on wtotal's 5. `level_by_city` opens it up per held-out city:
- **`level_err`** = predicted mean − observed mean log-rate (LOCO). Near 0 = the city is placed at the
  right height; a large value is the calibration miss the agency total is meant to fix.
- **`r2_oos`** = that city's own out-of-sample fit. A single strongly-negative city is what drags the
  pooled number down.


In [10]:
# Decompose each config's pooled r2_oos per held-out city on the top rung (ours+approved+agency).
CFG = "ours+approved+agency"
for cat in ["wtotal", "wprop"]:
    # One target at a time; Ridge then GBM so the level_err / r2_oos columns sit side by side.
    for model in ["ridge", "gbm"]:
        print(f"== {cat} · {CFG} · {model} · LOCO (worst-calibrated city first) ==")
        print(level_by_city(cat, CFG, model=model).to_string())
        print()

== wtotal · ours+approved+agency · ridge · LOCO (worst-calibrated city first) ==
                n  obs_mean  pred_mean  level_err  r2_oos
city                                                     
atlanta       369      3.41       5.28       1.87  -1.990
detroit       600      4.20       4.55       0.35  -0.020
houston      1536      3.45       3.59       0.15   0.336
chicago      2064      3.44       3.10      -0.34   0.367
kansas_city   445      3.74       3.57      -0.18   0.559

== wtotal · ours+approved+agency · gbm · LOCO (worst-calibrated city first) ==
                n  obs_mean  pred_mean  level_err  r2_oos
city                                                     
detroit       600      4.20       4.49       0.29   0.088
atlanta       369      3.41       3.99       0.57   0.202
houston      1536      3.45       3.28      -0.16   0.288
chicago      2064      3.44       3.59       0.15   0.421
kansas_city   445      3.74       3.44      -0.30   0.489

== wprop · ours+approved+a

**Read — per-city breakdown (why the pooled `r2_oos` moves).** The pooled number is one fit over all
held-out blocks, so one miscalibrated city dominates it.

- **wtotal / 5 — Ridge is sunk by a single city.** Atlanta `level_err` **+1.87** (predicts mean 5.28
  vs observed 3.41) gives it a per-city `r2_oos` of **−1.99**, dragging Ridge's pooled LOCO `r2_oos` to
  **0.194** while the other four cities sit at 0.34–0.56. GBM has no blow-up (Atlanta `level_err` +0.57,
  `r2_oos` 0.20) → pooled **0.375**. This lone outlier — not a general skill gap — is the whole wtotal
  calibration story. Do **not** tune toward cross-city `r2_oos` on n = 5.
- **wprop / 9 — with the agency total, Ridge is well-calibrated everywhere.** Every city's `level_err`
  is within ±0.17 and every per-city `r2_oos` is positive (0.03–0.47) → pooled **0.380**. GBM instead
  over-predicts Atlanta (**+0.54** → `r2_oos` 0.09) and San Francisco (+0.28 → 0.11) → pooled **0.359**.
  The linear model's even calibration is what wins on the wider target.
- **Rank stays healthy for both regardless of the calibration wobble** — judge extrapolation on the
  9-city property evidence, not wtotal/5.


### 4·b — Why some held-out cities fit well and others don't

The pooled `r2_oos` hides two *different* ways a held-out city can fit badly. This opens every LOCO city
on four diagnostics:

- **`obs_sd`** — spread of observed block-group log-rates *within* the city. This is the variance the
  model gets to explain; a city with little internal contrast has a small denominator (SST), which caps
  `r2_oos` no matter how good the predictions are.
- **`within_corr`** — Pearson of predicted vs observed *inside* the city (does the ranking bite there).
- **`level_err`** — predicted minus observed city-mean log-rate: the calibration / leveling miss.
- **`feat_dist`** — RMS z-distance of the city's predictor centroid from the pooled centre: how
  *atypical* the city looks in feature space, i.e. how far LOCO must extrapolate to place it.


In [11]:
# Per-held-out-city fit diagnostics. The ladder runs above already RETAIN each fold's fitted model
# (training.run_cv keeps them), so these read the same cached runs instead of refitting.
TOP = "ours+approved+agency"          # the deployed candidate rung

# wtotal is LightGBM's home (5 cities); wprop is Ridge's (9). Worst-fitting city first.
for cat, model in [("wtotal", "gbm"), ("wprop", "ridge")]:
    print(f"== {cat} \u00b7 {model} \u00b7 LOCO \u2014 why a held-out city fits or not (worst r2_oos first) ==")
    print(diagnostics.city_diag(get_run(cat, TOP, "loco", model), _POOL[cat]).to_string())
    print()


== wtotal · gbm · LOCO — why a held-out city fits or not (worst r2_oos first) ==


                n  obs_mean  obs_sd  level_err  r2_oos  within_corr  feat_dist
city                                                                          
detroit       600      4.20    0.80       0.29   0.088         0.47       0.60
atlanta       369      3.41    1.18       0.57   0.202         0.66       1.03
houston      1536      3.45    1.18      -0.16   0.288         0.58       0.28
chicago      2064      3.44    1.05       0.15   0.421         0.67       0.27
kansas_city   445      3.74    1.27      -0.30   0.489         0.75       0.50

== wprop · ridge · LOCO — why a held-out city fits or not (worst r2_oos first) ==


                  n  obs_mean  obs_sd  level_err  r2_oos  within_corr  feat_dist
city                                                                            
detroit         600      3.95    0.69       0.14   0.034         0.37       0.64
san_francisco   634      3.28    0.92       0.10   0.196         0.48       0.61
pittsburgh      300      3.06    1.00       0.11   0.288         0.55       0.35
houston        1536      3.21    1.10       0.07   0.303         0.56       0.28
chicago        2064      3.20    0.80      -0.09   0.340         0.60       0.30
sacramento      360      2.92    0.93       0.03   0.363         0.61       0.43
atlanta         369      3.19    1.00      -0.00   0.372         0.61       0.96
jacksonville    538      2.67    1.04       0.06   0.411         0.65       0.54
kansas_city     445      3.57    1.12      -0.17   0.473         0.72       0.47



**Read — two failure modes, and why Kansas City is the easy one.**

- **Detroit fits worst because it has the least to explain, not because the model is wrong.** It is a
  uniformly high-crime city — the highest mean (`obs_mean` **4.20**) but the *narrowest* within-city
  spread (`obs_sd` **0.80**) and the weakest internal signal (`within_corr` **0.47**). A small spread is a
  tiny SST, so even sensible predictions score a low `r2_oos` (**0.088**). Property tells the same story
  (`obs_sd` 0.69 → `r2_oos` 0.03): Detroit is a floor set by its own flatness.
- **Atlanta fits badly for the opposite reason — it is the feature-space outlier.** Its spread is
  healthy (`obs_sd` 1.18) but `feat_dist` is by far the largest (**1.03** vs 0.3–0.6 elsewhere) and the
  5-city model over-levels it (`level_err` **+0.57**) → `r2_oos` **0.202**. Give it more neighbours and a
  calibrated level and the miss vanishes: on **wprop / 9** Atlanta's `level_err` is **0.00** and its
  `r2_oos` jumps to **0.372**. Atlanta is a *sample-size / extrapolation* problem, not a data problem.
- **Kansas City fits best because it is the ideal held-out city: wide spread, typical features.** The
  *widest* within-city range (`obs_sd` **1.27**), the *strongest* internal signal (`within_corr` **0.75**),
  and a middling `feat_dist` (0.50) → `r2_oos` **0.489** (and **0.473** on property). Plenty of genuine
  block-to-block variation for in-distribution predictors to grab.

**This points at the same lever as §4/§6: more cities.** Detroit's ceiling is intrinsic (a flat city
stays flat), but Atlanta's miss is pure extrapolation — every added city shrinks the `feat_dist` an
unseen city can have, which is exactly what turns Atlanta-on-5 (0.20) into Atlanta-on-9 (0.37).


### 5 · Hyperparameter tuning — does Optuna move LightGBM?

The GBM above uses fixed defaults. Here we let **Optuna** search LightGBM's hyperparameters, tuning
**directly on the LOCO `r2_oos`** — the leave-one-city-out number that reflects the national rollout,
the metric we actually want to maximise.

- **Objective = LOCO `r2_oos`.** We optimise the deployment metric itself rather than an 80/20 proxy,
  so any lift lands on the axis that matters for unseen cities.
- **Honesty caveat.** With only 9 cities, the reported best LOCO `r2_oos` is the *max over 40 trials*
  on the same 9-fold structure, so it carries some selection optimism — read the magnitude as an
  upper bound, and check the per-city breakdown that the lift isn't carried by a single city. The
  **80/20** column below is the untouched cross-check.


In [12]:
# Tune LightGBM on the PROPERTY target (9 cities), the wider/less-brittle sample.
TUNE_CAT, TUNE_PSET = "wprop", "ours+approved+agency"


def _gbm_pooled(params, split, field):
    """Fit tuned-GBM through the same training orchestrator; return one POOLED metric."""
    run = fit_run(TUNE_CAT, TUNE_PSET, split, "gbm", gbm_params=params)
    return float(fold_metrics(run).loc["POOLED"][field])


def _objective(trial):
    """Maximise LOCO r2_oos directly -- the leave-one-city-out (extrapolation) deployment metric."""
    params = dict(
        n_estimators=trial.suggest_int("n_estimators", 200, 1200, step=100),
        learning_rate=trial.suggest_float("learning_rate", 0.01, 0.1, log=True),
        num_leaves=trial.suggest_int("num_leaves", 8, 127),
        min_child_samples=trial.suggest_int("min_child_samples", 10, 150),
        subsample=trial.suggest_float("subsample", 0.6, 1.0),
        colsample_bytree=trial.suggest_float("colsample_bytree", 0.6, 1.0),
        reg_alpha=trial.suggest_float("reg_alpha", 0.0, 5.0),
        reg_lambda=trial.suggest_float("reg_lambda", 0.0, 5.0),
    )
    return _gbm_pooled(params, "loco", "r2_oos")


# The 40-trial LOCO study is deterministic (seed=42) but slow (~15 min: LOCO refits 9 folds per trial).
# Flip RUN_OPTUNA to re-run it live; otherwise use the cached best params from that seeded study.
RUN_OPTUNA = False
if RUN_OPTUNA:
    study = optuna.create_study(direction="maximize", sampler=optuna.samplers.TPESampler(seed=42))
    study.optimize(_objective, n_trials=40, show_progress_bar=False)
    best = study.best_params
    print("best LOCO r2_oos:", round(study.best_value, 4))
else:
    # Cached optimum of the seeded 40-trial LOCO study (best LOCO r2_oos = 0.378).
    best = dict(n_estimators=700, learning_rate=0.0131, num_leaves=122, min_child_samples=105,
                subsample=0.9981, colsample_bytree=0.9378, reg_alpha=4.1183, reg_lambda=4.4776)
    print(f"best LOCO r2_oos: {_gbm_pooled(best, 'loco', 'r2_oos'):.4f}  "
          "(cached seed=42 40-trial optimum; set RUN_OPTUNA=True to re-run the full search)")
print("best params:", {k: (round(v, 4) if isinstance(v, float) else v) for k, v in best.items()})


def _scorecard(params, tag):
    """Print LOCO + 80/20 calibration/rank for a param set through the same pipeline."""
    for split in ["loco", "80_20"]:
        run = fit_run(TUNE_CAT, TUNE_PSET, split, "gbm", gbm_params=params)
        m = fold_metrics(run).loc["POOLED"]
        rk = rank_metrics(run, TUNE_CAT)
        print(f"  {tag:8s} {split:5s} | r2_oos={m['r2_oos']:.3f} mae={m['mae']:.3f} "
              f"recall@10={rk['recall@10']:.3f} recall@25={rk['recall@25']:.3f} "
              f"called_safe={rk['called_safe']:.3f}")


print("\nwprop · default vs Optuna-tuned LightGBM (objective = LOCO r2_oos; 80/20 is the cross-check):")
_scorecard(None, "default")
_scorecard(best, "tuned")

best LOCO r2_oos: 0.3730  (cached seed=42 40-trial optimum; set RUN_OPTUNA=True to re-run the full search)
best params: {'n_estimators': 700, 'learning_rate': 0.0131, 'num_leaves': 122, 'min_child_samples': 105, 'subsample': 0.9981, 'colsample_bytree': 0.9378, 'reg_alpha': 4.1183, 'reg_lambda': 4.4776}

wprop · default vs Optuna-tuned LightGBM (objective = LOCO r2_oos; 80/20 is the cross-check):


  default  loco  | r2_oos=0.359 mae=0.590 recall@10=0.367 recall@25=0.513 called_safe=0.199


  default  80_20 | r2_oos=0.490 mae=0.520 recall@10=0.390 recall@25=0.557 called_safe=0.157


  tuned    loco  | r2_oos=0.373 mae=0.580 recall@10=0.381 recall@25=0.522 called_safe=0.190


  tuned    80_20 | r2_oos=0.474 mae=0.520 recall@10=0.411 recall@25=0.559 called_safe=0.162


**Read — tuning.** Optuna optimises the LOCO `r2_oos` directly and buys a **modest, honest lift**:

- LOCO `r2_oos` **0.359 → 0.373** (+0.014), `mae` 0.590 → 0.580, `recall@10` 0.367 → 0.381,
  `recall@25` 0.513 → 0.522, `called_safe` 0.199 → 0.190 — every metric nudges the right way. It is essentially the same ceiling
  (~0.38) the 80/20-tuned model reached, so tuning *toward* LOCO doesn't unlock more than the proxy did.
- **It's genuine extrapolation, not interpolation.** The tuned model's untouched **80/20** `r2_oos`
  actually slips (0.490 → 0.474) — Optuna traded a little in-sample fit for the LOCO gain, exactly what
  we want.
- **Verdict: a real but second-order lever.** Best params ask for heavy regularization (slow
  `learning_rate` ≈ 0.013, strong `min_child_samples` 105 / L1–L2 ≈ 4 — Optuna wants *less*
  overfitting). Worth keeping, but small next to the two first-order levers: adding the **agency
  predictor** and adding **cities**.


### 6 · Which predictors are doing the work?

Two honest reads of predictor value, each on its model's home target:

- **`coef_mean` / `sign_stab` (Ridge · wprop / 9)** — the standardized per-SD effect and how often the
  9 LOCO folds agree on its sign. Direction + reliability.
- **`gain_share` (GBM · wtotal / 5)** — share of tree split-gain the feature captured (how much the
  trees *used* it).
- **`perm_dr2` (both)** — the out-of-sample test: pooled `r2_oos` lost when the feature is shuffled
  inside each held-out city. This is the "does it actually help predict an *unseen* city" number.

`perm_dr2` only scores *within-city-varying* predictors; the city-constant LEVEL features (agency
anchor, division dummies) shuffle to zero, so their value is read from the ladder (§2) and the coef.


In [13]:
# Feature value two ways: in-model use (Ridge coef / GBM gain) and out-of-sample worth (LOCO
# permutation). Reuses the fold models retained on the cached runs.
run_r = get_run("wprop", TOP, "loco", "ridge")
base_r, perm_r = diagnostics.perm_importance(run_r, _POOL["wprop"])
tbl_r = diagnostics.ridge_coef_table(run_r).join(perm_r).sort_values("perm_dr2", ascending=False)
print(f"RIDGE \u00b7 wprop / 9 cities \u00b7 {TOP}   (base LOCO r2_oos = {base_r:.3f})")
print("  coef_mean = per-SD effect \u00b7 sign_stab = fold agreement \u00b7 perm_dr2 = pooled r2_oos lost when shuffled")
print(tbl_r.to_string())
print()

run_g = get_run("wtotal", TOP, "loco", "gbm")
base_g, perm_g = diagnostics.perm_importance(run_g, _POOL["wtotal"])
tbl_g = pd.DataFrame({"gain_share": diagnostics.gbm_gain(run_g)}).join(perm_g).sort_values("gain_share", ascending=False)
print(f"GBM \u00b7 wtotal / 5 cities \u00b7 {TOP}   (base LOCO r2_oos = {base_g:.3f})")
print("  gain_share = share of tree split-gain \u00b7 perm_dr2 = pooled r2_oos lost when shuffled")
print(tbl_g.to_string())
print("\n* perm_dr2 is NaN for city-constant LEVEL features (agency anchor, division dummies): shuffling")
print("  within one held-out city can't move them. Their worth is the ladder LEVEL lift (\u00a72) + the Ridge coef.")


RIDGE · wprop / 9 cities · ours+approved+agency   (base LOCO r2_oos = 0.380)
  coef_mean = per-SD effect · sign_stab = fold agreement · perm_dr2 = pooled r2_oos lost when shuffled
                                coef_mean  sign_stab  perm_dr2
transit_service_intensity_logc      0.167       1.00    0.0576
transit_has_transit                 0.158       1.00    0.0438
vacant_pct_log                      0.127       1.00    0.0318
vacant_pct_lag6_log                 0.130       1.00    0.0272
clip_foreclosure_pct_lag6_log       0.123       1.00    0.0235
in_household_pct                    0.098       1.00    0.0196
own_pct                            -0.101       1.00    0.0192
unq_gas_stations_clips              0.081       1.00    0.0124
det_pct                            -0.081       1.00    0.0092
hardscapes_pct_avg                 -0.076       1.00    0.0079
unq_convenience_stores_clips        0.062       1.00    0.0078
moved1yr_pct                        0.073       1.00    0.0077
p

GBM · wtotal / 5 cities · ours+approved+agency   (base LOCO r2_oos = 0.375)
  gain_share = share of tree split-gain · perm_dr2 = pooled r2_oos lost when shuffled
                                gain_share  perm_dr2
vacant_pct_lag6_log                 0.1703    0.0428
vacant_pct_log                      0.1493    0.0449
transit_service_intensity_logc      0.1090    0.0994
own_pct                             0.0899    0.0484
clip_foreclosure_pct_lag6_log       0.0589    0.0477
transit_nearest_stop_m_log          0.0449    0.0183
pop_est_5mile_log                   0.0342    0.0016
roof_condition_avg                  0.0333    0.0056
hardscapes_pct_avg                  0.0314    0.0064
moved1yr_pct                        0.0314    0.0187
pop_ch_1mile                        0.0301    0.0014
det_pct                             0.0286    0.0328
clip_foreclosure_pct_log            0.0215    0.0196
lap_pct                             0.0207    0.0043
transit_has_transit                 0.0196 

**Read — the workhorses are shared across models; imagery and POI counts are marginal.**

- **The same four feature families carry BOTH models.** Top by out-of-sample `perm_dr2`:
  **transit service intensity** (Ridge 0.058 / GBM 0.099 — the single biggest OOS lever in each),
  **vacancy** (current + 6-month lag, ~0.03–0.05 each), **lagged foreclosure** (~0.024 / 0.048), and
  **tenure & occupancy** (`own_pct`, `in_household_pct`, ~0.02). Distress + transit exposure + who lives
  there is the signal, and it is sign-stable across every fold (`sign_stab` = 1.0).
- **Imagery and POI counts are close to dead weight — the natural next-iteration cut.** `roof_*`
  (debris, condition, missing-material), `hardscapes`, liquor / gas-station counts, `lap_pct` and
  `transit_overnight_stop_share` all sit at |`perm_dr2`| ≲ 0.005, and several flip sign across folds
  (`roof_debris` `sign_stab` 0.44, `lap_pct` 0.56) — they add variance, not skill. GBM spends real
  split-gain on `roof_` / `hardscapes` (in-sample) yet they earn almost nothing out-of-sample
  (`perm_dr2` ≈ 0) — the classic in-sample-vs-LOCO gap.
- **The agency anchor's value is real but invisible here by construction.** Its `perm_dr2` is NaN
  (city-constant), yet it is the decisive LEVEL predictor from §2 (Ridge coef **0.126**, sign-stable;
  ladder `r2_oos` 0.36 → 0.38). Level and within-city are different jobs: the anchor does the first, the
  transit / vacancy / tenure block does the second.
- **Agreement between a linear and a tree model is the reassuring part** — the ranking of *what matters*
  barely changes between RidgeCV and LightGBM, so the signal lives in the features, not in a modelling
  artifact. That is the shortlist to defend, and the weak tail is where feature pruning and the Next-steps predictor brainstorm should focus.


### Next steps

1. **More cities — the highest-leverage move.** Every number here rests on 5–9 cities, and LOCO is
   harsh precisely because each held-out city is structurally unique. Adding POC cities shrinks the
   80/20→LOCO gap and de-fragilizes the cross-city `r2_oos` far more than any modelling change.
2. **Tuning (section 5).** Optuna buys a modest, honest lift on LOCO calibration — worth keeping, but
   it is a second-order lever behind sample size and features. Do **not** tune toward the cross-city
   `r2_oos` on this few cities.
3. **Predictors — target *within-city* signal.** The agency lag already carries the between-city level;
   new features should explain block-to-block variation (the incumbent's weak spot and Detroit's):
   finer POI mix, local built-environment / imagery, temporal crime lags where available, and
   mobility / foot-traffic. Brainstorm and screen these the same way — LOCO first, RANK metrics first.
